cpp/daal/include/services/internal/sycl/buffer_utils_sycl.h

```diff
@@ -83,23 +128,71 @@ class BufferCopier
             DAAL_ASSERT_UNIVERSAL_BUFFER_TYPE(srcUnivers, T);
             DAAL_ASSERT_UNIVERSAL_BUFFER_TYPE(dstUnivers, T);
 
-            auto src = srcUnivers.get<T>().toSycl(status);
-            DAAL_CHECK_STATUS_RETURN_VOID_IF_FAIL(status);
+            const auto & src_buffer = srcUnivers.get<T>();
+            const auto & dst_buffer = dstUnivers.get<T>();
+
+            DAAL_ASSERT(src_buffer.size() >= srcOffset + count);
+            DAAL_ASSERT(dst_buffer.size() >= dstOffset + count);
+
+#ifdef DAAL_SYCL_INTERFACE_USM
+            if (src_buffer.isUSMBacked() && dst_buffer.isUSMBacked())
+            {
+                auto src = src_buffer.toUSM(status);
+                DAAL_CHECK_STATUS_RETURN_VOID_IF_FAIL(status);
+
+                auto dst = dst_buffer.toUSM(status);
+                DAAL_CHECK_STATUS_RETURN_VOID_IF_FAIL(status);
 
-            auto dst = dstUnivers.get<T>().toSycl(status);
-            DAAL_CHECK_STATUS_RETURN_VOID_IF_FAIL(status);
+                auto src_raw = src.get() + srcOffset;
+                auto dst_raw = dst.get() + dstOffset;
 
-            DAAL_ASSERT(src.get_count() >= srcOffset + count);
-            DAAL_ASSERT(dst.get_count() >= dstOffset + count);
+                const size_t bytes_count = sizeof(T) * count;
+                DAAL_ASSERT(bytes_count >= count);
```

Comment:
I suppose it is a redundant check since `sizeof` returns `size_t` that is non-negative by definition. Thus this check can be done in compile-time `static_assert(sizeof(T) != 0);`

Reply:
it's an overflow check